packages/arb-bridge-eth/contracts/arch/Marshaling.sol

```diff
@@ -0,0 +1,255 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+pragma solidity ^0.5.11;
+
+import "./Value.sol";
+import "./Hashing.sol";
+
+import "../libraries/BytesLib.sol";
+
+library Marshaling {
+    using BytesLib for bytes;
+    using Value for Value.Data;
+
+    function deserializeHashPreImage(bytes memory data, uint256 startOffset)
+        internal
+        pure
+        returns (
+            uint256, // offset
+            Value.Data memory
+        )
+    {
+        uint256 size;
+        require(
+            data.length >= startOffset && data.length - startOffset >= 64,
+            "to short"
+        );
+        bytes32 hashData = data.toBytes32(startOffset);
+        startOffset += 32;
+        (startOffset, size) = deserializeInt(data, startOffset);
+        return (startOffset, Value.newTuplePreImage(hashData, size));
+    }
+
+    function deserializeInt(bytes memory data, uint256 startOffset)
+        internal
+        pure
+        returns (
+            uint256, // offset
+            uint256 // val
+        )
+    {
+        require(
+            data.length >= startOffset && data.length - startOffset >= 32,
+            "too short"
+        );
+        return (startOffset + 32, data.toUint(startOffset));
+    }
+
+    function deserializeCheckedInt(bytes memory data, uint256 startOffset)
+        internal
+        pure
+        returns (
+            bool, // valid
+            uint256, // offset
+            uint256 // val
+        )
+    {
+        uint256 totalLength = data.length;
+        if (
+            totalLength < startOffset ||
+            totalLength - startOffset < 33 ||
+            uint8(data[startOffset]) != Value.intTypeCode()
+        ) {
+            return (false, startOffset, 0);
+        }
+        return (true, startOffset + 33, data.toUint(startOffset + 1));
+    }
+
+    function deserializeCodePoint(bytes memory data, uint256 startOffset)
+        internal
+        pure
+        returns (
+            uint256, // offset
+            Value.Data memory // val
+        )
+    {
+        uint256 offset = startOffset;
+        uint8 immediateType = uint8(data[offset]);
+        offset++;
+        uint8 opCode = uint8(data[offset]);
+        offset++;
+        Value.Data memory immediate;
+        if (immediateType == 1) {
+            (offset, immediate) = deserialize(data, offset);
+        }
+        bytes32 nextHash = data.toBytes32(offset);
+        offset += 32;
+        if (immediateType == 1) {
+            return (offset, Value.newCodePoint(opCode, nextHash, immediate));
+        }
+        return (offset, Value.newCodePoint(opCode, nextHash));
+    }
```

Comment:
perhaps some helper functions that returns the values from `data` and the correct offset? more surface area for bugs having to manually increment offset in different locations everytime

Reply:
Agreed. I added some more helper functions so that we now don't have to manually mess with the offset in any of the more complex functions